src/viam/sdk/services/mlmodel/client.cpp

```diff
@@ -45,77 +39,36 @@ std::shared_ptr<MLModelService::named_tensor_views> MLModelServiceClient::infer(
     auto* const resp = pb::Arena::CreateMessage<mlpb::InferResponse>(arena.get());
     grpc::ClientContext ctx;
 
-    if (!kUseFlatTensors) {
-        struct tensor_storage_and_views {
-            mlmodel_details::tensor_storage storage;
-            MLModelService::named_tensor_views views;
-        };
-        auto tsav = std::make_shared<tensor_storage_and_views>();
-
-        auto& mutable_input_data = *req->mutable_input_data();
-        auto& mutable_input_data_fields = *mutable_input_data.mutable_fields();
-
-        // TODO: Currently, this doesn't validate that we are passing the
-        // right input type in. We could query the metadata here and
-        // consult it.
-        for (const auto& kv : inputs) {
-            pb::Value& value = mutable_input_data_fields[kv.first];
-            mlmodel_details::tensor_to_pb_value(kv.second, &value);
-        }
-
-        const auto result = stub_->Infer(&ctx, *req, resp);
-        if (!result.ok()) {
-            throw std::runtime_error(result.error_message());
-        }
-
-        // TODO(RSDK-3298): This is an extra RPC on every inference, but
-        // it is not clear that caching it is safe.
-        const auto md = metadata();
-
-        const auto& output_fields = resp->output_data().fields();
-        for (const auto& output : md.outputs) {
-            const auto where = output_fields.find(output.name);
-            // Ignore any outputs for which we don't have metadata, since
-            // we can't know what type they should decode to.
-            if (where != output_fields.end()) {
-                mlmodel_details::pb_value_to_tensor(
```

Comment:
Ditto `mlmodel_details::pb_value_to_tensor` and associated helpers.

Reply:
Can do.